File: csharp/VIPDiscountHandler.cs
namespace ConsoleApp1
{
    public class VIPDiscountHandler : DiscountHandler
    {
        public override decimal CalculateDiscount(Customer customer, decimal orderTotal)
        {
            if (customer.IsVIP)
            {
                return orderTotal * 0.8m; // 20% discount
            }

            return this.nextHandler?.CalculateDiscount(customer, orderTotal) ?? orderTotal;
        }
    }
}

File: ts/vip-discount.handler.ts
import { DiscountHandler } from "./calculate-discount";
import { Customer } from "./customer";

export class VIPDiscountHandler extends DiscountHandler {
  calculateDiscount(customer: Customer, orderTotal: number): number {
    if (customer.isVIP) {
      return orderTotal * 0.8; // 20% discount
    }

    return (
      this.nextHandler?.calculateDiscount(customer, orderTotal) ?? orderTotal
    );
  }
}
